ddtrace/llmobs/_utils.py

```diff
@@ -17,59 +21,154 @@
 from ddtrace.llmobs._constants import OPENAI_APM_SPAN_NAME
 from ddtrace.llmobs._constants import SESSION_ID
 from ddtrace.llmobs._constants import VERTEXAI_APM_SPAN_NAME
+from ddtrace.llmobs.utils import Message
+from ddtrace.llmobs.utils import Prompt
 from ddtrace.trace import Span
 
 
 log = get_logger(__name__)
 
-
-def validate_prompt(prompt: dict) -> Dict[str, Union[str, dict, List[str]]]:
-    validated_prompt = {}  # type: Dict[str, Union[str, dict, List[str]]]
+# SemVer regex from https://semver.org/
+SEMVER_PATTERN_COMPILED = compile(
+    r"^(?P<major>0|[1-9]\d*)\."
+    r"(?P<minor>0|[1-9]\d*)\."
+    r"(?P<patch>0|[1-9]\d*)"
+    r"(?:-(?P<prerelease>(?:0|[1-9]\d*|\d*[a-zA-Z-]"
+    r"[0-9a-zA-Z-]*)(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-]"
+    r"[0-9a-zA-Z-]*))*))?"
+    r"(?:\+(?P<buildmetadata>[0-9a-zA-Z-]+"
+    r"(?:\.[0-9a-zA-Z-]+)*))?$"
+)
+
+PromptDict = Dict[str, Union[str, Dict[str, Any], List[str], List[Dict[str, str]], List[Message]]]
+
+
+def _validate_prompt(
+    prompt: Union[Dict[str, Any], Prompt], ml_app: str = "", strict_validation: bool = True
+) -> PromptDict:
+    # Stage 0: Check if dict
     if not isinstance(prompt, dict):
-        raise TypeError("Prompt must be a dictionary")
+        raise TypeError(f"Prompt must be a dictionary, got {type(prompt).__name__}.")
+
+    # Stage 1: Extract values
+    name = prompt.get("name")
+    prompt_id = prompt.get("id")
+    version = prompt.get("version")
     variables = prompt.get("variables")
     template = prompt.get("template")
-    version = prompt.get("version")
-    prompt_id = prompt.get("id")
+    chat_template = prompt.get("chat_template")
     ctx_variable_keys = prompt.get("rag_context_variables")
-    rag_query_variable_keys = prompt.get("rag_query_variables")
+    query_variable_keys = prompt.get("rag_query_variables")
+
+    # Stage 2: Strict validations
+    if strict_validation:
+        _strict_validate_prompt(prompt)
+
+    # Stage 3: Set defaults
+    final_prompt_id = prompt_id or name or DEFAULT_PROMPT_NAME
+    final_name = name or prompt_id or DEFAULT_PROMPT_NAME
+    final_version = version or "1.0.0"
+    final_ctx_variable_keys = ctx_variable_keys or ["context"]
+    final_query_variable_keys = query_variable_keys or ["question"]
+
+    # Stage 4: Type checks
+    if not isinstance(final_prompt_id, str):
+        raise TypeError(f"'id' must be str, got {type(final_prompt_id).__name__}.")
+
+    if not isinstance(final_name, str):
+        raise TypeError(f"'name' must be str, got {type(final_name).__name__}.")
+
+    if not isinstance(final_version, str):
+        raise TypeError(f"'version' must be str, got {type(final_version).__name__}.")
+
+    if not (isinstance(final_ctx_variable_keys, list) and all(isinstance(i, str) for i in final_ctx_variable_keys)):
```

Comment:
could probably make a `is_str_list_instance` method since we check this below too